File: androidApp/src/main/java/com/tourtrip/android/ui/tours/ToursViewModel.kt
package com.tourtrip.android.ui.tours

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tourtrip.shared.data.remote.ApiClient
import com.tourtrip.shared.data.repository.TourRepository
import com.tourtrip.shared.domain.models.Tour
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for Tours Screen
 */
class ToursViewModel : ViewModel() {

    private val apiClient = ApiClient()
    private val tourRepository = TourRepository(apiClient)

    private val _uiState = MutableStateFlow<ToursUiState>(ToursUiState.Loading)
    val uiState: StateFlow<ToursUiState> = _uiState.asStateFlow()

    init {
        loadTours()
    }

    fun loadTours() {
        viewModelScope.launch {
            _uiState.value = ToursUiState.Loading

            tourRepository.getTours().fold(
                onSuccess = { tours ->
                    _uiState.value = ToursUiState.Success(tours)
                },
                onFailure = { error ->
                    _uiState.value = ToursUiState.Error(error.message ?: "Unknown error")
                }
            )
        }
    }

    fun searchTours(query: String) {
        viewModelScope.launch {
            _uiState.value = ToursUiState.Loading

            tourRepository.searchTours(query).fold(
                onSuccess = { tours ->
                    _uiState.value = ToursUiState.Success(tours)
                },
                onFailure = { error ->
                    _uiState.value = ToursUiState.Error(error.message ?: "Search failed")
                }
            )
        }
    }

    override fun onCleared() {
        super.onCleared()
        apiClient.close()
    }
}

/**
 * UI State for Tours Screen
 */
sealed class ToursUiState {
    object Loading : ToursUiState()
    data class Success(val tours: List<Tour>) : ToursUiState()
    data class Error(val message: String) : ToursUiState()
}

File: iosApp/iosApp/ViewModels/ToursViewModel.swift
import Foundation
import Combine
import shared

/**
 * Tours ViewModel (iOS)
 * Manages tours list state using Combine
 */
class ToursViewModel: ObservableObject {
    @Published var state: ToursState = .loading

    private let apiClient = ApiClient(baseUrl: "https://tourtrip.app/api")
    private var tourRepository: TourRepository?

    init() {
        tourRepository = TourRepository(apiClient: apiClient)
    }

    func loadTours() {
        state = .loading

        // Using KMM shared repository
        Task {
            do {
                if let repository = tourRepository {
                    let result = try await repository.getTours(
                        category: nil,
                        city: nil,
                        minPrice: nil,
                        maxPrice: nil
                    )

                    await MainActor.run {
                        if let tours = result as? [Tour] {
                            state = .success(tours)
                        } else {
                            state = .error("Failed to load tours")
                        }
                    }
                }
            } catch {
                await MainActor.run {
                    state = .error(error.localizedDescription)
                }
            }
        }
    }

    func searchTours(query: String) {
        state = .loading

        Task {
            do {
                if let repository = tourRepository {
                    let result = try await repository.searchTours(query: query)

                    await MainActor.run {
                        if let tours = result as? [Tour] {
                            state = .success(tours)
                        } else {
                            state = .error("Search failed")
                        }
                    }
                }
            } catch {
                await MainActor.run {
                    state = .error(error.localizedDescription)
                }
            }
        }
    }

    deinit {
        apiClient.close()
    }
}

/**
 * UI State for Tours
 */
enum ToursState {
    case loading
    case success([Tour])
    case error(String)
}
